src/shogun/multiclass/BruteSolver.cpp

```diff
@@ -0,0 +1,68 @@
+#include <shogun/multiclass/KNNSolver.h>
+#include <shogun/lib/Signal.h>
+
+using namespace shogun;
+
+CBRUTESolver::CBRUTESolver(int32_t k, float64_t q, int32_t num_classes, int32_t min_label, SGVector<int32_t> train_labels, SGMatrix<index_t> NN)
+{
+	init();
+
+	m_k=k;
+	m_q=q;
+	m_num_classes=num_classes;
+	m_min_label=min_label;
+	m_train_labels=train_labels;
+	nn=NN;
+}
+
+void CBRUTESolver::init()
+{
+	m_k=3;
+	m_q=1.0;
+	m_num_classes=0;
+	m_min_label=0;
+	m_train_labels=0;
+	nn=SGMatrix<index_t>(3, 0);
+
+}
+
+CMulticlassLabels* CBRUTESolver::classify_objects(CDistance* distance, int32_t num_lab, int32_t* train_lab, float64_t* classes)
+{
+	CMulticlassLabels* output=new CMulticlassLabels(num_lab);
+	//get the k nearest neighbors of each example
+	SGMatrix<index_t> NN = this->nn;
+
+	//from the indices to the nearest neighbors, compute the class labels
+	for (int32_t i=0; i<num_lab && (!CSignal::cancel_computations()); i++)
+	{
+		//write the labels of the k nearest neighbors from theirs indices
+		for (int32_t j=0; j<m_k; j++)
+			train_lab[j] = m_train_labels[ NN(j,i) ];
+
+		//get the index of the 'nearest' class
+		int32_t out_idx = choose_class(classes, train_lab);
+		//write the label of 'nearest' in the output
+		output->set_label(i, out_idx + m_min_label);
+	}
+
+	return output;
+}
+
+int32_t* CBRUTESolver:: classify_objects_k(CDistance* distance, int32_t num_lab, int32_t* train_lab, int32_t* classes)
+{
+	int32_t* output=SG_MALLOC(int32_t, m_k*num_lab);
+
+	//get the k nearest neighbors of each example
+	SGMatrix<index_t> NN = this->nn;
+
+	for (int32_t i=0; i<num_lab && (!CSignal::cancel_computations()); i++)
```

Comment:
All these nested for loops in all of the `classify_objects_k` is:
 * essentially is the same across all the solvers, they are just depending on different inputs
 * really ugly; please think about refactoring it to use matrix/vector operations, that way the operator can be optimised later for parallel execution.